### PR TITLE
backward compatibility fix on edge case

### PR DESCRIPTION
**Pull request by sasurobert**
backward compatibility fix on edge case.

Current code running on mainnet, does save account even if execution fails later. This means the account will get an empty hash instead of a NIL one. This results in roothash missmatch. This PR fixes the issue.

### CODE REVIEW

File: genesis/process/genesisBlockCreator.go [hunk 1]
```diff
@@ -417,6 +418,7 @@ func (gbc *genesisBlockCreator) computeDNSAddresses() error {
 		DataPool:           gbc.arg.Data.Datapool(),
 		CompiledSCPool:     gbc.arg.Data.Datapool().SmartContracts(),
 		NilCompiledSCStore: true,
+		EpochNotifier:      forking.NewGenericEpochNotifier(),
```
Comment (iulianpascalau) [inline genesis/process/genesisBlockCreator.go:421]: \+ `SaveAccountsIfErrorDisableEpoch: unreachableEpoch`

File: genesis/process/metaGenesisBlockCreator.go [hunk 2]
```diff
@@ -252,15 +259,9 @@ func createProcessorsForMetaGenesisBlock(arg ArgsGenesisBlockCreator, enableEpoc
 		DataPool:           arg.Data.Datapool(),
 		CompiledSCPool:     arg.Data.Datapool().SmartContracts(),
 		NilCompiledSCStore: true,
+		EpochNotifier:      epochNotifier,
```
Comment (iulianpascalau) [inline genesis/process/metaGenesisBlockCreator.go:262]: \+ `SaveAccountsIfErrorDisableEpoch: enableEpochs.SaveAccountsIfErrorDisableEpoch`
and add in `genesis/process/shardGenesisBlockCreator.go` createGenesisConfig function, `SaveAccountsIfErrorDisableEpoch: unreachableEpoch`

File: genesis/process/shardGenesisBlockCreator.go [hunk 3]
```diff
@@ -298,6 +298,7 @@ func createProcessorsForShardGenesisBlock(arg ArgsGenesisBlockCreator, enableEpo
 		DataPool:           arg.Data.Datapool(),
 		CompiledSCPool:     arg.Data.Datapool().SmartContracts(),
 		NilCompiledSCStore: true,
+		EpochNotifier:      epochNotifier,
```
Comment (iulianpascalau) [inline genesis/process/shardGenesisBlockCreator.go:301]: same thing as in metaGenesisBlockCreator.go

File: process/smartContract/hooks/blockChainHook.go [hunk 4]
```diff
@@ -630,6 +658,12 @@ func (bh *BlockChainHookImpl) RevertToSnapshot(snapshot int) error {
 	return bh.accounts.RevertToSnapshot(snapshot)
 }
 
+// EpochConfirmed is called whenever a new epoch is confirmed
+func (bh *BlockChainHookImpl) EpochConfirmed(epoch uint32, _ uint64) {
+	bh.flagSaveAccountsIfError.Toggle(epoch < bh.saveAccountsIfErrorDisableEpoch)
+	log.Debug("blockchain hook: save accounts if error", "enabled", bh.flagSaveAccountsIfError.IsSet())
```
Comment (iulianpascalau) [inline process/smartContract/hooks/blockChainHook.go:664]: log.Debug("blockchain hook: built in functions save accounts if error"...

**Review by SebastianMarian — DISMISSED** [review]

File: process/smartContract/hooks/blockChainHook.go (hunk 4 above)
Comment (sasurobert) [reply process/smartContract/hooks/blockChainHook.go:664]: done

File: genesis/process/genesisBlockCreator.go (hunk 1 above)
Comment (sasurobert) [reply genesis/process/genesisBlockCreator.go:421]: done, although not needed :D

File: genesis/process/metaGenesisBlockCreator.go (hunk 2 above)
Comment (sasurobert) [reply genesis/process/metaGenesisBlockCreator.go:262]: done, although not needed :D

File: genesis/process/shardGenesisBlockCreator.go (hunk 3 above)
Comment (sasurobert) [reply genesis/process/shardGenesisBlockCreator.go:301]: done, although not needed :D

File: genesis/process/metaGenesisBlockCreator.go (hunk 2 above)
Comment (iulianpascalau) [reply genesis/process/metaGenesisBlockCreator.go]: this is needed in hardfork.

**Review by iulianpascalau — DISMISSED** [review]

File: process/smartContract/hooks/blockChainHook.go [hunk 5]
```diff
@@ -661,7 +661,7 @@ func (bh *BlockChainHookImpl) RevertToSnapshot(snapshot int) error {
 // EpochConfirmed is called whenever a new epoch is confirmed
 func (bh *BlockChainHookImpl) EpochConfirmed(epoch uint32, _ uint64) {
 	bh.flagSaveAccountsIfError.Toggle(epoch < bh.saveAccountsIfErrorDisableEpoch)
-	log.Debug("blockchain hook: built in functions save accounts if error", "enabled", bh.flagSaveAccountsIfError.IsSet())
+	log.Debug("blockchain hook: built in functions save accounts always", "enabled", bh.flagSaveAccountsIfError.IsSet())
```
Comment (SebastianMarian) [inline process/smartContract/hooks/blockChainHook.go:664]: "if error" was ok, as this flag is used only to save in error case as long as it is enabled

Comment (sasurobert) [reply process/smartContract/hooks/blockChainHook.go:664]: no it was not - as it appeared at every epoch change on log.

**Pull request closed without merging**